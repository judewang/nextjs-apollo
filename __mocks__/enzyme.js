import _ from 'lodash';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

// https://github.com/airbnb/enzyme/issues/1295
export function diveTo(component, selector, options) {
  const result = component.find(selector);

  if (result.length > 0 || !component.dive) return result;
  const instance = component.instance() || {};

  const next = component.dive({
    ...options,
    context: _.assign(
      _.get(options, ['context'], {}),
      instance.getChildContext && instance.getChildContext(),
    ),
  });

  return diveTo(next, selector, options);
}

export * from 'enzyme';
