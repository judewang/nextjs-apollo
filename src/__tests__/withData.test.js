import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import http from 'http';
import gql from 'graphql-tag';
import express from 'express';
import ws from 'ws';
import bodyParser from 'body-parser';
import { PubSub } from 'graphql-subscriptions';
import { SubscriptionServer } from 'subscriptions-transport-ws';
import { render } from 'enzyme';
import { graphql } from 'react-apollo';
import { execute } from 'apollo-link';
import { graphqlExpress } from 'apollo-server-express';
import {
  subscribe, GraphQLString, GraphQLSchema, GraphQLObjectType,
} from 'graphql';
import { withData, initApollo } from '..';

const pubsub = new PubSub();

const resolveSpy = jest.fn(payload => payload * 2);
const subscribeSpy = jest.fn(() => pubsub.asyncIterator('me'));

const schema = new GraphQLSchema({
  query: new GraphQLObjectType({
    name: 'Query',
    fields: {
      me: {
        type: GraphQLString,
        resolve: resolveSpy,
      },
    },
  }),
  subscription: new GraphQLObjectType({
    name: 'Subscription',
    fields: {
      me: {
        type: GraphQLString,
        subscribe: subscribeSpy,
      },
    },
  }),
});

const app = express();
app.use('/graphql', bodyParser.json(), graphqlExpress(
  req => ({ schema, context: _.pick(req.headers, ['x-hackers']) }),
));

const server = http.createServer(app);
server.listen();

const { port } = server.address();

new SubscriptionServer({ // eslint-disable-line
  schema,
  subscribe,
  execute,
  onConnect: params => params,
}, { server, path: '/' });

class App extends React.Component {
  static propTypes = {
    value: PropTypes.number,
    data: PropTypes.shape(),
  }

  static defaultProps = {
    value: 0,
    data: {},
  }

  static async getInitialProps() {
    return { value: 99 };
  }

  render() {
    const { value, data } = this.props;

    return (
      <div>
        {`${_.get(data, ['me'])} value: ${value}`}
      </div>
    );
  }
}

const apollo = {
  httpUri: `http://localhost:${port}/graphql`,
  wsUri: `ws://localhost:${port}/`,
  webSocketImpl: ws,
};

describe('withData', () => {
  describe('process.browser is false', () => {
    beforeEach(() => {
      process.browser = false;
    });

    it('when NoSSR', async () => {
      const WrapApp = withData({ ...apollo, noSSR: true })(
        graphql(gql`query { me }`)(App),
      );

      const props = await WrapApp.getInitialProps();
      expect(props).toMatchSnapshot();
      expect(resolveSpy).toHaveBeenCalledTimes(0);

      const component = render(<WrapApp {...props} />);
      expect(component).toMatchSnapshot();
    });

    it('when getInitialProps', async () => {
      resolveSpy.mockReturnValueOnce(Promise.resolve('hello'));
      const WrapApp = withData(apollo)(
        graphql(gql`query { me }`)(App),
      );

      const props = await WrapApp.getInitialProps();
      expect(props).toMatchSnapshot();

      expect(resolveSpy).toHaveBeenLastCalledWith(
        undefined, {}, {}, expect.anything(),
      );
      expect(resolveSpy).toHaveBeenCalledTimes(1);

      const component = render(<WrapApp {...props} />);
      expect(component).toMatchSnapshot();
    });

    it('when no getInitialProps', async () => {
      resolveSpy.mockReturnValueOnce(Promise.resolve('hello'));
      const WrapApp = withData(apollo)(graphql(gql`query { me }`)(() => <div />));
      expect(await WrapApp.getInitialProps()).toMatchSnapshot();
      expect(resolveSpy).toHaveBeenCalledTimes(1);
    });

    it('when throw error', async () => {
      resolveSpy.mockImplementationOnce(() => Promise.reject(new Error()));
      const WrapApp = withData(apollo)(graphql(gql`query { me }`)(App));

      const props = await WrapApp.getInitialProps();
      expect(props).toMatchSnapshot();

      expect(resolveSpy).toHaveBeenLastCalledWith(
        undefined, {}, {}, expect.anything(),
      );
      expect(resolveSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe('process.browser is true', () => {
    beforeEach(() => {
      process.browser = true;
      initApollo.client = null;
    });

    it('getInitialProps', async () => {
      const WrapApp1 = withData(apollo)(graphql(gql`query { me }`)(App));
      expect(await WrapApp1.getInitialProps()).toEqual({ value: 99 });

      const WrapApp2 = withData(apollo)(graphql(gql`query { me }`)(App));
      expect(await WrapApp2.getInitialProps()).toEqual({ value: 99 });

      expect(resolveSpy).toHaveBeenCalledTimes(0);
    });

    it('no websocket', async () => {
      const WrapApp = withData({ ...apollo, wsUri: undefined })(
        graphql(gql`query { me }`)(App),
      );
      expect(await WrapApp.getInitialProps()).toEqual({ value: 99 });
      expect(resolveSpy).toHaveBeenCalledTimes(0);
    });
  });

  describe('initApollo', () => {
    beforeEach(() => {
      process.browser = true;
      initApollo.client = null;
    });

    it('init', async () => {
      expect(initApollo()).toMatchSnapshot();
      expect(initApollo()).toBe(initApollo());
    });

    it('subscription', async () => {
      let resolve;
      const promise = new Promise((__) => { resolve = __; });
      subscribeSpy.mockImplementationOnce(resolve);

      const client = initApollo({}, apollo);
      execute(client.link, { query: gql`subscription { me }` }).subscribe({});

      await promise;

      expect(subscribeSpy).toHaveBeenLastCalledWith(undefined, {}, {}, expect.anything());
      expect(subscribeSpy).toHaveBeenCalledTimes(1);
    });

    it('query', async () => {
      let resolve;
      const promise = new Promise((__) => { resolve = __; });
      resolveSpy.mockImplementationOnce(resolve);

      const client = initApollo({}, apollo);
      execute(client.link, { query: gql`query { me }` }).subscribe({});
      await promise;

      expect(resolveSpy).toHaveBeenLastCalledWith(undefined, {}, {}, expect.anything());
      expect(resolveSpy).toHaveBeenCalledTimes(1);
    });

    it('headers', async () => {
      let resolve;
      const promise = new Promise((__) => { resolve = __; });
      resolveSpy.mockImplementationOnce(resolve);

      const client = initApollo({}, { ...apollo, headers: () => ({ 'x-hackers': '999' }) });
      execute(client.link, { query: gql`query { me }` }).subscribe({});
      await promise;

      expect(resolveSpy).toHaveBeenLastCalledWith(undefined, {}, { 'x-hackers': '999' }, expect.anything());
      expect(resolveSpy).toHaveBeenCalledTimes(1);
    });
  });
});
