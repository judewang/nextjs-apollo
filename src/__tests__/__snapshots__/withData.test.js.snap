// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withData initApollo init 1`] = `
ApolloClient {
  "cache": InMemoryCache {
    "addTypename": true,
    "cacheKeyRoot": CacheKeyNode {
      "children": null,
      "key": null,
    },
    "config": Object {
      "addTypename": true,
      "cacheResolvers": undefined,
      "dataIdFromObject": undefined,
      "fragmentMatcher": IntrospectionFragmentMatcher {
        "isReady": false,
        "match": [Function],
      },
    },
    "data": DepTrackingCache {
      "data": Object {},
      "depend": [Function],
    },
    "maybeBroadcastWatch": [Function],
    "optimistic": Array [],
    "silenceBroadcast": false,
    "storeReader": StoreReader {
      "cacheKeyRoot": CacheKeyNode {
        "children": null,
        "key": null,
      },
      "executeSelectionSet": [Function],
      "executeStoreQuery": [Function],
      "keyMaker": QueryKeyMaker {
        "cacheKeyRoot": CacheKeyNode {
          "children": null,
          "key": null,
        },
        "perQueryKeyMakers": Map {},
      },
    },
    "storeWriter": StoreWriter {},
    "typenameDocumentCache": Map {},
    "watches": Set {},
  },
  "defaultOptions": Object {},
  "disableNetworkFetches": false,
  "link": ApolloLink {
    "request": [Function],
  },
  "mutate": [Function],
  "query": [Function],
  "queryDeduplication": true,
  "reFetchObservableQueries": [Function],
  "resetStore": [Function],
  "resetStoreCallbacks": Array [],
  "ssrMode": false,
  "store": DataStore {
    "cache": InMemoryCache {
      "addTypename": true,
      "cacheKeyRoot": CacheKeyNode {
        "children": null,
        "key": null,
      },
      "config": Object {
        "addTypename": true,
        "cacheResolvers": undefined,
        "dataIdFromObject": undefined,
        "fragmentMatcher": IntrospectionFragmentMatcher {
          "isReady": false,
          "match": [Function],
        },
      },
      "data": DepTrackingCache {
        "data": Object {},
        "depend": [Function],
      },
      "maybeBroadcastWatch": [Function],
      "optimistic": Array [],
      "silenceBroadcast": false,
      "storeReader": StoreReader {
        "cacheKeyRoot": CacheKeyNode {
          "children": null,
          "key": null,
        },
        "executeSelectionSet": [Function],
        "executeStoreQuery": [Function],
        "keyMaker": QueryKeyMaker {
          "cacheKeyRoot": CacheKeyNode {
            "children": null,
            "key": null,
          },
          "perQueryKeyMakers": Map {},
        },
      },
      "storeWriter": StoreWriter {},
      "typenameDocumentCache": Map {},
      "watches": Set {},
    },
  },
  "version": "2.4.2",
  "watchQuery": [Function],
}
`;

exports[`withData process.browser is false when NoSSR 1`] = `
Object {
  "value": 99,
}
`;

exports[`withData process.browser is false when NoSSR 2`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "data": "undefined value: 99",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`withData process.browser is false when getInitialProps 1`] = `
Object {
  "cache": Object {
    "ROOT_QUERY": Object {
      "me": "hello",
    },
  },
  "value": 99,
}
`;

exports[`withData process.browser is false when getInitialProps 2`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "data": "hello value: 99",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`withData process.browser is false when no getInitialProps 1`] = `
Object {
  "cache": Object {
    "ROOT_QUERY": Object {
      "me": "hello",
    },
  },
  "url": Object {},
}
`;

exports[`withData process.browser is false when throw error 1`] = `
Object {
  "value": 99,
}
`;
